### PR TITLE
Optimize nonpayable assertion

### PR DESCRIPTION
**Pull request by iamdefinitelyahuman**
### What I did
Optimize how the nonpayable check is applied within the bytecode

### How I did it
Where possible, external functions are now organized by [payable], [nonpayable] and a single assertion that `msg.value == 0` is placed in between the two sections.

When a contract contains a payable default function, and nonpayable functions, this optimization is not possible and the check is still applied per-function.

### How to verify it
Run the tests.  I expanded the testing around the `@payable` decorator to handle all the different possible permutations based on this optimization.

### Cute Animal Picture
image

### CODE REVIEW

File: vyper/parser/function_definitions/parse_external_function.py [hunk 1]
```diff
@@ -33,14 +33,13 @@ def get_external_arg_copier(
...
 def parse_external_function(
-    code: vy_ast.FunctionDef, sig: FunctionSignature, context: Context, is_contract_payable: bool
+    code: vy_ast.FunctionDef, sig: FunctionSignature, context: Context, check_nonpayable: bool,
 ) -> LLLnode:
     """
     Parse a external function (FuncDef), and produce full function body.
 
     :param sig: the FuntionSignature
     :param code: ast of function
-    :param is_contract_payable: bool - does this contract contain payable functions?
```
Comment (fubuloubu) [inline vyper/parser/function_definitions/parse_external_function.py:43]: missing `check_nonpayable` arg?

File: vyper/parser/parser.py [hunk 2]
```diff
@@ -86,20 +87,36 @@ def parse_external_interfaces(external_interfaces, global_ctx):
...
 
-def parse_other_functions(
-    o, otherfuncs, sigs, external_interfaces, global_ctx, default_function, is_contract_payable,
-):
+def parse_other_functions(o, otherfuncs, sigs, external_interfaces, global_ctx, default_function):
+    # check for payable/nonpayable external functions to optimize nonpayable assertions
+    func_types = [i._metadata["type"] for i in global_ctx._defs]
+    mutabilities = [i.mutability for i in func_types if i.visibility == FunctionVisibility.EXTERNAL]
+    has_payable = next((True for i in mutabilities if i == StateMutability.PAYABLE), False)
+    has_nonpayable = next((True for i in mutabilities if i != StateMutability.PAYABLE), False)
```
Comment (fubuloubu) [inline vyper/parser/parser.py:95]: Seems clearer that this be a straight `bool`
```suggestion
    has_payable = next((True for i in mutabilities if i == StateMutability.PAYABLE), None) is None
    has_nonpayable = next((True for i in mutabilities if i != StateMutability.PAYABLE), None) is None
```

File: vyper/parser/parser.py [hunk 3]
```diff
@@ -86,20 +87,36 @@ def parse_external_interfaces(external_interfaces, global_ctx):
...
+        func_type = func_node._metadata["type"]
         func_lll = parse_function(
-            func_node, {**{"self": sigs}, **external_interfaces}, global_ctx, is_contract_payable,
+            func_node, {**{"self": sigs}, **external_interfaces}, global_ctx, check_per_function
         )
-        if func_lll.context.is_internal:
+        if func_type.visibility == FunctionVisibility.INTERNAL:
             internal_func_sub.append(func_lll)
+        elif func_type.mutability == StateMutability.PAYABLE:
+            add_gas += 30
```
Comment (fubuloubu) [inline vyper/parser/parser.py:118]: random `30` lol

**Review by fubuloubu — APPROVED** [review]

File: vyper/parser/parser.py (hunk 3 above)
Comment (iamdefinitelyahuman) [reply vyper/parser/parser.py:118]: this is related to the gas estimation.  we plan to remove it, but that seemed outside of the scope of this (and more work).  but yeah i agree it's weird, and hopefully will not be there.

**Pull request merged**